cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,1126 @@
+public:
+  __device__ __forceinline__ void ConsumeTile(BufferOffsetT tile_id)
+  {
+    // Offset into this tile's buffers
+    BufferOffsetT buffer_offset = tile_id * BUFFERS_PER_BLOCK;
+
+    // Indicates whether all of this tiles items are within bounds
+    bool is_full_tile = buffer_offset + BUFFERS_PER_BLOCK < num_buffers;
+
+    // Load the buffer sizes of this tile's buffers
+    BufferSizeIteratorT tile_buffer_sizes_it = buffer_sizes_it + buffer_offset;
+    BufferSizeT buffer_sizes[BUFFERS_PER_THREAD];
+    if (is_full_tile)
+    {
+      LoadBufferSizesFullTile(tile_buffer_sizes_it, buffer_sizes);
+    }
+    else
+    {
+      LoadBufferSizesPartialTile(tile_buffer_sizes_it, buffer_sizes, num_buffers - buffer_offset);
+    }
+
+    // Ensure we can repurpose the BlockLoad's temporary storage
+    CTA_SYNC();
+
+    // Count how many buffers fall into each size-class
+    VectorizedSizeClassCounterT size_class_histogram = {};
+    GetBufferSizeClassHistogram(buffer_sizes, size_class_histogram);
+
+    // Compute the prefix sum over the histogram
+    VectorizedSizeClassCounterT size_class_agg = {};
+    BlockSizeClassScanT(temp_storage.size_scan_storage)
+      .ExclusiveSum(size_class_histogram, size_class_histogram, size_class_agg);
+
+    // Ensure we can repurpose the scan's temporary storage for scattering the buffer ids
+    CTA_SYNC();
+
+    // Factor in the per-size-class counts / offsets
+    // That is, WLEV buffer offset has to be offset by the TLEV buffer count and BLEV buffer offset
+    // has to be offset by the TLEV+WLEV buffer count
+    uint32_t buffer_count = 0U;
+    for (uint32_t i = 0; i < NUM_SIZE_CLASSES; i++)
+    {
+      size_class_histogram.Add(i, buffer_count);
+      buffer_count += size_class_agg.Get(i);
+    }
+
+    // Signal the number of BLEV buffers we're planning to write out
+    BufferOffsetT buffer_exclusive_prefix = 0;
+    if (tile_id == 0)
+    {
+      if (threadIdx.x == 0)
+      {
+        blev_buffer_scan_state.SetInclusive(tile_id, size_class_agg.Get(BLEV_SIZE_CLASS));
+      }
+      buffer_exclusive_prefix = 0;
+    }
+    else
+    {
+      BLevBuffScanPrefixCallbackOpT blev_buffer_prefix_op(blev_buffer_scan_state,
+                                                          temp_storage.buffer_scan_callback,
+                                                          Sum(),
+                                                          tile_id);
+
+      // Signal our partial prefix and wait for the inclusive prefix of previous tiles
+      if (threadIdx.x < CUB_PTX_WARP_THREADS)
+      {
+        buffer_exclusive_prefix = blev_buffer_prefix_op(size_class_agg.Get(BLEV_SIZE_CLASS));
+      }
+    }
+    if (threadIdx.x == 0)
+    {
+      temp_storage.blev_buffer_offset = buffer_exclusive_prefix;
+    }
+
+    // Ensure the prefix callback has finished using its temporary storage and that it can be reused
+    // in the next stage
+    CTA_SYNC();
```

Comment:
`temp_storage.blev_buffer_offset` is not used in `PartitionBuffersBySize`. Since look-back is rather expensive, do you think there's any advantage in overlapping of decoupled look-back with the `PartitionBuffersBySize` in other warps? `BLevBuffScanPrefixCallbackOpT ` storage should be like 4 ints, so putting it into struct instead of a union shouldn't increase shared memory requirements significantly, but we would get rid of one sync and overlap some operations.